File: crates/pray-core/src/destination.rs
use crate::manifest::{
    DestinationEntry, DestinationMode, ExportRole, Manifest, ManifestLocal, ManifestPackage,
    ManifestTarget,
};
use crate::{PrayError, PrayResult};

pub fn is_local_path_form(value: &str) -> bool {
    value.starts_with('.')
        || value.starts_with('/')
        || value.ends_with(".md")
        || value.ends_with(".txt")
        || value.ends_with(".markdown")
}

pub fn destination_target_name(mode: DestinationMode, path: &str) -> String {
    let prefix = match mode {
        DestinationMode::Compose => "compose",
        DestinationMode::Tree => "tree",
        DestinationMode::Legacy => "legacy",
    };
    format!("{prefix}:{path}")
}

pub fn new_destination_target(mode: DestinationMode, path: &str) -> ManifestTarget {
    let mut target = ManifestTarget {
        name: destination_target_name(mode, path),
        mode,
        scoped: true,
        ..ManifestTarget::default()
    };
    match mode {
        DestinationMode::Compose => target.outputs.push(path.to_string()),
        DestinationMode::Tree => target.skills.push(path.to_string()),
        DestinationMode::Legacy => {}
    }
    target
}

pub fn upsert_package(manifest: &mut Manifest, package: ManifestPackage) -> PrayResult<()> {
    if let Some(existing) = manifest
        .packages
        .iter_mut()
        .find(|candidate| candidate.name == package.name)
    {
        if existing.constraint != package.constraint
            && existing.constraint != "*"
            && package.constraint != "*"
        {
            return Err(PrayError::Manifest(format!(
                "package {} declared with conflicting constraints ({} vs {})",
                package.name, existing.constraint, package.constraint
            )));
        }
        if existing.constraint == "*" && package.constraint != "*" {
            existing.constraint = package.constraint;
        }
        if existing.source.is_none() {
            existing.source = package.source;
        } else if package.source.is_some() && existing.source != package.source {
            return Err(PrayError::Manifest(format!(
                "package {} declared with conflicting sources",
                package.name
            )));
        }
        for export in package.exports {
            if !existing.exports.contains(&export) {
                existing.exports.push(export);
            }
        }
        for role in package.roles {
            if !existing.roles.contains(&role) {
                existing.roles.push(role);
            }
        }
        if package.file.is_some() {
            if existing.file.is_some() && existing.file != package.file {
                return Err(PrayError::Manifest(format!(
                    "package {} declared with conflicting file: destinations",
                    package.name
                )));
            }
            existing.file = package.file;
        }
        existing.bound = existing.bound || package.bound;
        existing.optional = existing.optional || package.optional;
        if existing.path.is_none() {
            existing.path = package.path;
        }
        if existing.git.is_none() {
            existing.git = package.git;
        }
        if existing.tag.is_none() {
            existing.tag = package.tag;
        }
        if existing.rev.is_none() {
            existing.rev = package.rev;
        }
        if existing.tarball.is_none() {
            existing.tarball = package.tarball;
        }
        if existing.oci.is_none() {
            existing.oci = package.oci;
        }
        for group in package.groups {
            if !existing.groups.contains(&group) {
                existing.groups.push(group);
            }
        }
        return Ok(());
    }
    manifest.packages.push(package);
    Ok(())
}

pub fn upsert_local(manifest: &mut Manifest, local: ManifestLocal) {
    if let Some(existing) = manifest
        .local
        .iter_mut()
        .find(|candidate| candidate.path == local.path)
    {
        existing.bound = existing.bound || local.bound;
        existing.optional = existing.optional || local.optional;
        if existing.position == "after" && local.position != "after" {
            existing.position = local.position;
        }
        if existing.file.is_none() {
            existing.file = local.file;
        }
        return;
    }
    manifest.local.push(local);
}

pub fn bind_package_entry(target: &mut ManifestTarget, package_name: &str) {
    let entry = DestinationEntry::Package {
        name: package_name.to_string(),
    };
    if !target.entries.contains(&entry) {
        target.entries.push(entry);
    }
}

pub fn bind_local_entry(target: &mut ManifestTarget, path: &str) {
    let entry = DestinationEntry::Local {
        path: path.to_string(),
    };
    if !target.entries.contains(&entry) {
        target.entries.push(entry);
    }
}

pub fn role_for_destination(mode: DestinationMode) -> Option<ExportRole> {
    match mode {
        DestinationMode::Compose => Some(ExportRole::Fragment),
        DestinationMode::Tree => Some(ExportRole::Folder),
        DestinationMode::Legacy => None,
    }
}

pub fn package_bound_to_compose(package: &ManifestPackage, target: &ManifestTarget) -> bool {
    if target.scoped && target.mode == DestinationMode::Compose {
        return target.entries.iter().any(|entry| match entry {
            DestinationEntry::Package { name } => name == &package.name,
            DestinationEntry::Local { .. } => false,
        });
    }
    if package.bound || package.file.is_some() {
        return false;
    }
    true
}

pub fn package_bound_to_tree(package: &ManifestPackage, target: &ManifestTarget) -> bool {
    if target.scoped && target.mode == DestinationMode::Tree {
        return target.entries.iter().any(|entry| match entry {
            DestinationEntry::Package { name } => name == &package.name,
            DestinationEntry::Local { .. } => false,
        });
    }
    if package.bound || package.file.is_some() {
        return false;
    }
    true
}

pub fn export_kind_matches_role(kind: &str, role: ExportRole) -> bool {
    match role {
        ExportRole::Fragment => kind == "fragment",
        ExportRole::Folder => matches!(kind, "folder" | "skill"),
        ExportRole::File => kind == "file",
    }
}

pub fn local_is_compose_embed(manifest: &Manifest, local: &ManifestLocal) -> bool {
    if local.file.is_some() {
        return false;
    }
    !manifest.targets.iter().any(|target| {
        target.mode == DestinationMode::Tree
            && target.entries.iter().any(|entry| match entry {
                DestinationEntry::Local { path } => path == &local.path,
                DestinationEntry::Package { .. } => false,
            })
    })
}

pub fn try_apply_local_pray(
    manifest: &mut Manifest,
    path: &str,
    file_dest: Option<&str>,
    other_package_signal: bool,
    destination_index: Option<usize>,
) -> PrayResult<bool> {
    if !is_local_path_form(path) {
        return Ok(false);
    }
    if file_dest.is_some() {
        if other_package_signal {
            return Ok(false);
        }
        return Err(PrayError::Parse {
            kind: "manifest",
            message: "a local file cannot use file:".to_string(),
        });
    }
    if other_package_signal {
        return Ok(false);
    }
    let Some(index) = destination_index else {
        return Err(PrayError::Parse {
            kind: "manifest",
            message: "local pray paths are only valid inside compose".to_string(),
        });
    };
    let mode = manifest.targets[index].mode;
    if mode != DestinationMode::Compose {
        return Err(PrayError::Parse {
            kind: "manifest",
            message: "local pray paths are only valid inside compose".to_string(),
        });
    }
    let local = ManifestLocal {
        path: path.to_string(),
        position: "after".to_string(),
        optional: false,
        bound: true,
        file: None,
    };
    bind_local_entry(&mut manifest.targets[index], &local.path);
    upsert_local(manifest, local);
    Ok(true)
}

File: rubygems/pray-cli/lib/pray/destination.rb
# frozen_string_literal: true

module Pray
  DestinationEntry = Struct.new(:kind, :name, :path) do
    def self.package(name)
      new(kind: "package", name: name, path: nil)
    end

    def self.local(path)
      new(kind: "local", name: nil, path: path)
    end
  end

  module Destination
    module_function

    def local_path_form?(value)
      value.start_with?(".", "/") ||
        value.end_with?(".md", ".txt", ".markdown")
    end

    def destination_target_name(mode, path)
      prefix = case mode
      when "compose" then "compose"
      when "tree" then "tree"
      else "legacy"
      end
      "#{prefix}:#{path}"
    end

    def new_destination_target(mode, path)
      target = ManifestTarget.new(
        name: destination_target_name(mode, path),
        mode: mode,
        scoped: true
      )
      case mode
      when "compose" then target.outputs << path
      when "tree" then target.skills << path
      end
      target
    end

    def upsert_package(manifest, package)
      existing = manifest.packages.find { |candidate| candidate.name == package.name }
      unless existing
        manifest.packages << package
        return
      end

      if existing.constraint != package.constraint &&
          existing.constraint != "*" &&
          package.constraint != "*"
        raise Error.manifest(
          "package #{package.name} declared with conflicting constraints " \
          "(#{existing.constraint} vs #{package.constraint})"
        )
      end
      existing.constraint = package.constraint if existing.constraint == "*" && package.constraint != "*"

      if existing.source.nil?
        existing.source = package.source
      elsif package.source && existing.source != package.source
        raise Error.manifest("package #{package.name} declared with conflicting sources")
      end

      package.exports.each do |export|
        existing.exports << export unless existing.exports.include?(export)
      end
      package.roles.each do |role|
        existing.roles << role unless existing.roles.include?(role)
      end

      if package.file
        if existing.file && existing.file != package.file
          raise Error.manifest(
            "package #{package.name} declared with conflicting file: destinations"
          )
        end
        existing.file = package.file
      end

      existing.bound = existing.bound || package.bound
      existing.optional = existing.optional || package.optional
      existing.path ||= package.path
      existing.git ||= package.git
      existing.tag ||= package.tag
      existing.rev ||= package.rev
      existing.tarball ||= package.tarball
      existing.oci ||= package.oci
      package.groups.each do |group|
        existing.groups << group unless existing.groups.include?(group)
      end
    end

    def upsert_local(manifest, local)
      existing = manifest.local.find { |candidate| candidate.path == local.path }
      unless existing
        manifest.local << local
        return
      end

      existing.bound = existing.bound || local.bound
      existing.optional = existing.optional || local.optional
      if existing.position == "after" && local.position != "after"
        existing.position = local.position
      end
      existing.file ||= local.file
    end

    def bind_package_entry(target, package_name)
      entry = DestinationEntry.package(package_name)
      return if target.entries.any? { |candidate| candidate.kind == "package" && candidate.name == package_name }

      target.entries << entry
    end

    def bind_local_entry(target, path)
      return if target.entries.any? { |candidate| candidate.kind == "local" && candidate.path == path }

      target.entries << DestinationEntry.local(path)
    end

    def role_for_destination(mode)
      case mode
      when "compose" then "fragment"
      when "tree" then "folder"
      end
    end

    def package_bound_to_compose?(package, target)
      if target.scoped && target.mode == "compose"
        return target.entries.any? { |entry| entry.kind == "package" && entry.name == package.name }
      end
      return false if package.bound || package.file

      true
    end

    def package_bound_to_tree?(package, target)
      if target.scoped && target.mode == "tree"
        return target.entries.any? { |entry| entry.kind == "package" && entry.name == package.name }
      end
      return false if package.bound || package.file

      true
    end

    def export_kind_matches_role?(kind, role)
      case role
      when "fragment" then kind == "fragment"
      when "folder" then %w[folder skill].include?(kind)
      when "file" then kind == "file"
      else false
      end
    end

    def local_compose_embed?(manifest, local)
      return false if local.file

      manifest.targets.none? do |target|
        target.mode == "tree" &&
          target.entries.any? { |entry| entry.kind == "local" && entry.path == local.path }
      end
    end

    def try_apply_local_pray(manifest, path, file_dest, other_package_signal, destination_index)
      return false unless local_path_form?(path)

      if file_dest
        return false if other_package_signal
        raise Error.parse("manifest", "a local file cannot use file:")
      end
      return false if other_package_signal
      unless destination_index
        raise Error.parse("manifest", "local pray paths are only valid inside compose")
      end
      mode = manifest.targets[destination_index].mode
      unless mode == "compose"
        raise Error.parse("manifest", "local pray paths are only valid inside compose")
      end
      local = ManifestLocal.new(path: path, bound: true)
      bind_local_entry(manifest.targets[destination_index], local.path)
      upsert_local(manifest, local)
      true
    end
  end
end
